### PR TITLE
BigInteger implementation

### PR DESCRIPTION
**Pull request by ikornienko**
This is a Pull Request to review @vicky17d's implementation of BigInteger.

### CODE REVIEW

File: src/main/java/implementations2/MyBigInteger.java [hunk 1]
```diff
@@ -0,0 +1,415 @@
...
+        return auxResultArray;
+    }
+
+
+//    private static int getSizeOfLongerArray(int[] array1, int[] array2) {
+//        return array1.length >= array2.length ? array1.length : array2.length;
+//    }
+
+
+    private static void validNumberCheck(String input) {
```
Comment (ikornienko) [inline src/main/java/implementations2/MyBigInteger.java]: In Java method names are usually verbs, so name ``checkNumber`` or something similar works better here.

File: src/main/java/implementations2/MyBigInteger.java [hunk 2]
```diff
@@ -0,0 +1,415 @@
...
+        String result = getUnsignedString(this.inputString);
+        result = trimLeadingZeroes(result);
+        if(result.length() == 0) return "0";
+
+        if(this.isSignNegative()) return "-" + result;
+        return result;
+
+    }
+
+    public boolean equals(MyBigInteger big2) {
```
Comment (ikornienko) [inline src/main/java/implementations2/MyBigInteger.java]: ``equals`` is a method of ``Object`` class, therefore it should be overridden properly with ``Object`` as an argument (and ``@Override`` annotation should be in place). It's important to have this method correctly overridden because Java compiler decides which overloaded method will be called (i.e. it's base on compile-time type), consider this code:

```java
MyBigInteger bg1 = new MyBigInteger("1");
Object bg2 = new MyBigInteger("1");
System.out.println(bg1.equals(bg2));  // prints "false"
```
I believe the intent was that these two should be equal, but because compile-time type of ``bg2`` is ``Object``, this method isn't even invoked. Such problem may hurt very badly when you'll be using ``MyBigInteger`` with collections (like keys for Map).

See EJ2 Item8 ("Effective Java 2nd edition", Item 8) for more details about ``equals`` contract.

File: src/main/java/implementations2/MyBigInteger.java [hunk 3]
```diff
@@ -0,0 +1,415 @@
...
+        if (this == big2) {
+            return true;
+        } else {
+            //check for sign and value
+            char sign1 = this.getSign();
+            char sign2 = big2.getSign();
+            String unsignedString1 = getUnsignedString(this.inputString);
+            String unsignedString2 = getUnsignedString(big2.inputString);
+            if ((sign1 == sign2) && unsignedString1.equals(unsignedString2)) {
```
Comment (ikornienko) [inline src/main/java/implementations2/MyBigInteger.java]: Just curious about "+0" and "-0", is it intentional that they are different? ;)

File: src/main/java/implementations2/MyBigInteger.java [hunk 4]
```diff
@@ -0,0 +1,415 @@
...
+                return true;
+            }
+        }
+        return false;
+
+    }
+
+    public int hashCode() {
+        Character sign = getSign();
```
Comment (ikornienko) [inline src/main/java/implementations2/MyBigInteger.java]: Actually unnecessary boxing here (which slightly can affect performance). Consider using ``char`` as is, since you always can case it to ``int`` and use in the expression for hash code calculation.

File: src/main/java/implementations2/MyBigInteger.java [hunk 5]
```diff
@@ -0,0 +1,415 @@
...
+    }
+
+    @Override
+    public int compareTo(MyBigInteger big2) {
+        if (this.equals(big2)) return 0;
+        if (bothArePositive(this, big2)) return compareMagnitude(big2);
+        if (bothAreNegative(this, big2)) return compareMagnitude(big2) * (-1);
+        if (this.isSignPositive()) return +1;
+        if (this.isSignNegative()) return -1;
+        return 0;  //this is not required.. check : Last statement could be above stmt without if i.e. return -1; but this is more readable code
```
Comment (ikornienko) [inline src/main/java/implementations2/MyBigInteger.java]: Then throw ``IllegalStateException`` here, even without any comment every developer would understand, that the intention that it never riches this statement (i.e. everything should happen above). Also it could save in case of "bad" refactoring (e.g. something was changed in the lines above or in the methods which is called from the lines above, and the intention is broken, then it'll start fail with ``IllegalStateException`` instead of silently mark two ``MyBigInteger`` as equal (``return 0;``).

File: src/main/java/implementations2/MyBigInteger.java [hunk 6]
```diff
@@ -0,0 +1,415 @@
...
+        } else if (str1.length() == str2.length()) {
+            for (int i = 0; i < str1.length(); i++) {
+                if (Character.getNumericValue(str1.charAt(i)) > Character.getNumericValue(str2.charAt(i))) {
+                    return +1;
+                } else if (Character.getNumericValue(str1.charAt(i)) < Character.getNumericValue(str2.charAt(i))) {
+                    return -1;
+                }
+            }
+        }
+        return 0;
```
Comment (ikornienko) [inline src/main/java/implementations2/MyBigInteger.java]: It's better to have ``throw new IllegalStateException();`` (same intention as in ``compareTo`` method).

File: src/main/java/implementations2/MyBigInteger.java [hunk 7]
```diff
@@ -0,0 +1,415 @@
...
+        } else if (str1.length() < str2.length()) {
+            return -1;
+        } else if (str1.length() == str2.length()) {
+            for (int i = 0; i < str1.length(); i++) {
+                if (Character.getNumericValue(str1.charAt(i)) > Character.getNumericValue(str2.charAt(i))) {
+                    return +1;
+                } else if (Character.getNumericValue(str1.charAt(i)) < Character.getNumericValue(str2.charAt(i))) {
+                    return -1;
+                }
+            }
```
Comment (ikornienko) [inline src/main/java/implementations2/MyBigInteger.java]: Better to move ``return 0;`` after this ``for`` statement (see comment below).

File: src/main/java/implementations2/MyBigInteger.java [hunk 8]
```diff
@@ -0,0 +1,415 @@
...
+        }
+        return 0;
+    }
+
+
+    public final char getSign() {
+        String str = this.inputString;
+        if (str.length() != 0) {
```
Comment (ikornienko) [inline src/main/java/implementations2/MyBigInteger.java]: Why do you do this check for ``str`` length? It seems that ``inputString`` could be never be an empty string (from implementation of ``validNumberCheck``).

File: src/main/java/implementations2/MyBigInteger.java [hunk 9]
```diff
@@ -0,0 +1,415 @@
...
+            char numOneFirstChar = str.charAt(0);
+            if ((numOneFirstChar == '+') || (numOneFirstChar == '-')) {
+                return numOneFirstChar;
+            }
+        }
+        return '+';
+    }
+
+    public final boolean isSignPositive() {
+        String str = this.inputString;
```
Comment (ikornienko) [inline src/main/java/implementations2/MyBigInteger.java]: Why not ``return getSign() == '+';``?

File: src/main/java/implementations2/MyBigInteger.java [hunk 10]
```diff
@@ -0,0 +1,415 @@
...
+        }
+
+        return str.substring(zeroesCount, str.length());
+    }
+
+
+    //todo change all methods to return MyBigInteger rather than String
+    public String add(MyBigInteger big2) {
+        char resultSign = '+';
+        int[] value1 = this.calculateInputStringToReversedArray();
```
Comment (ikornienko) [inline src/main/java/implementations2/MyBigInteger.java]: Here you've probably already had an idea: "Converting every time String to array of ints.. it's a performance issue right here! Maybe it's better to store array of ints+sign in the first place instead of String?" ;)

File: src/main/java/implementations2/MyBigInteger.java [hunk 11]
```diff
@@ -0,0 +1,415 @@
...
+        int zeroesCount = 0;
+        for (int i = 0; i < str.length(); i++) {
+            if (str.charAt(i) == '0') {
+                zeroesCount++;
+            } else {
+                break;
+            }
+        }
+
+        return str.substring(zeroesCount, str.length());
```
Comment (ikornienko) [inline src/main/java/implementations2/MyBigInteger.java]: It probably works (looking at the rest of the logic), but in case of "0" this method will return an empty string, which isn't what its name says (it trims leading zeroes, but "0" has no leading zeroes strictly speaking).

File: src/main/java/implementations2/MyBigInteger.java [hunk 12]
```diff
@@ -0,0 +1,415 @@
...
+        int[] resultArray = {};
+        if (bothArePositive(this, big2) ) {
+            resultArray = addAux(value1, value2);
+            resultSign = '+';
+        } else if (bothAreNegative(this, big2)){
+            resultArray = addAux(value1, value2);
+            resultSign = '-';
+        } else if(this.isSignPositive()) {  //todo - may be negate the number and send it to subtract
+//            resultArray = subtractAux()//
+            return this.subtract(big2);
```
Comment (ikornienko) [inline src/main/java/implementations2/MyBigInteger.java]: You probably want to substract (-1) * big2 (otherwise you'll add it, i.e. have a test "1 + (-1)" which likely will produce "2" instead of "0").

File: src/main/java/implementations2/MyBigInteger.java [hunk 13]
```diff
@@ -0,0 +1,415 @@
...
+            resultBuilder.deleteCharAt(resultBuilder.length() - 1);
+        }
+        if(resultChar == '-') {
+            resultBuilder.append('-');
+        }
+        return resultBuilder.reverse().toString();
+    }
+
+
+    private static int[] addAux(int[] array1, int[] array2) {
```
Comment (ikornienko) [inline src/main/java/implementations2/MyBigInteger.java]: Well done exercise here! (no sarcasm :) )

File: src/main/java/implementations2/MyBigInteger.java [hunk 14]
```diff
@@ -0,0 +1,415 @@
...
+
+    public String subtract(MyBigInteger big2) {
+        if (inputString.equals(big2.inputString)) {
+            return "0";
+        }
+        int[] value1 = this.calculateInputStringToReversedArray();
+        int[] value2 = big2.calculateInputStringToReversedArray();
+
+        int[] resultArray = {};
+        char resultSign = '+';
```
Comment (ikornienko) [inline src/main/java/implementations2/MyBigInteger.java]: For better readability consider not relying on any default value, and define value for ``resultSign`` in every ``if`` case you have.

File: src/main/java/implementations2/MyBigInteger.java [hunk 15]
```diff
@@ -0,0 +1,415 @@
...
+        if (bothArePositive(this, big2)) {
+            resultArray = subtractAux(value1, value2);
+            resultSign = '+';
+        }  else if (bothAreNegative(this, big2)) {
+            resultArray = subtractAux(value1, value2);
```
Comment (ikornienko) [inline src/main/java/implementations2/MyBigInteger.java]: It looks like substracting both negative values always gives ``resultSign`` "+", which isn't true...

File: src/test/java/implementations2/MyBigIntegerTest.java [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+import static org.junit.Assert.*;
+
+public class MyBigIntegerTest {
+
+    MyBigInteger big1 = new MyBigInteger("-333333333333333333333333");
```
Comment (ikornienko) [inline src/test/java/implementations2/MyBigIntegerTest.java]: It's really rare case when in tests you share something between tests, and it's usually not something that actually is UNDER tests. I.e. what if I want to adjust some case because I discovered that it adds no value, and should be replaced with another, if I change something here, it'll affect not only one test, but the whole bunch. I.e. it makes tests implicitly dependent, since they share the common data, therefore it makes things more complex. Just keep in mind, that opposite to the production code, it's ok to have a lot of duplication in test code (when this duplication isn't the behavior that should be extracted to some kind of a test framework :) ).

File: src/test/java/implementations2/MyBigIntegerTest.java [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+
+    }
+
+    @After
+    public void tearDown() throws Exception {
+
+    }
+
+    @Test
+    public void testToString() throws Exception {
```
Comment (ikornienko) [inline src/test/java/implementations2/MyBigIntegerTest.java]: Well done here with covering different cases!

File: src/test/java/implementations2/MyBigIntegerTest.java [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+        assertEquals("0", new MyBigInteger("+00000000000").toString());
+        assertEquals("0", new MyBigInteger("-0000000").toString());
+        assertEquals("-123456789123456789123456789", new MyBigInteger("-123456789123456789123456789").toString());
+        assertEquals("0", new MyBigInteger("-0").toString());
+
+    }
+
+    @Test
+    public void testEquals() throws Exception {
+        assertTrue(new MyBigInteger("123").equals(new MyBigInteger("123")));
```
Comment (ikornienko) [inline src/test/java/implementations2/MyBigIntegerTest.java]: I've already mentioned some corner cases with leading zeros and "+0"/"-0" that you may want to add here.

File: src/test/java/implementations2/MyBigIntegerTest.java [hunk 19]
```diff
@@ -0,0 +1,135 @@
...
+        assertTrue(new MyBigInteger("123").equals(new MyBigInteger("+123")));
+        assertTrue(new MyBigInteger("+123").equals(new MyBigInteger("123")));
+        assertTrue(new MyBigInteger("-123").equals(new MyBigInteger("-123")));
+        assertFalse(new MyBigInteger("-123").equals(new MyBigInteger("123")));
+    }
+
+    @Test
+    public void testHashCode() throws Exception {
+        assertEquals(new MyBigInteger("123").hashCode(), new MyBigInteger("123").hashCode());
```
Comment (ikornienko) [inline src/test/java/implementations2/MyBigIntegerTest.java]: Not that the test does something useful for you here... Usually you test ``hashCode`` and ``equals`` together (e.g.as described [here](http://stackoverflow.com/questions/4449728/how-can-i-do-unit-test-hashcode-function)). At least it's more clear for readers of the code what kind of contract you're verifying with tests.

File: src/test/java/implementations2/MyBigIntegerTest.java [hunk 20]
```diff
@@ -0,0 +1,135 @@
...
+        assertEquals(-1, (new MyBigInteger("-333333333333333333333")).compareTo(new MyBigInteger("333333333333333333334")));
+        assertEquals(-1, (new MyBigInteger("-333333333333333333333")).compareTo(new MyBigInteger("+333333333333333333333")));
+        assertEquals(1, (new MyBigInteger("333333333333333333333")).compareTo(new MyBigInteger("-333333333333333333334")));
+        assertEquals(-1, (new MyBigInteger("333333333333333333333")).compareTo(new MyBigInteger("433333333333333333333")));
+
+//        ArrayList listExpected = {-1234, -1123, -1, 10, 123, 1100};
+    }
+
+    @Test
+    public void testSorting() throws Exception {
```
Comment (ikornienko) [inline src/test/java/implementations2/MyBigIntegerTest.java]: It's ok to have it here, though it feels like not belonging to this class. If you have tests that validate ``Comparable`` contract, then this one is doing nothing (except verifying ``Collections.sort`` behavior).

**Comment by ikornienko** [comment]
Well done! There are some improvements that could be made, also the code style isn't very consistent (spaces etc.), but in overall I think it's a great start! Keep it coming :)

File: src/main/java/implementations2/MyBigInteger.java (hunk 2 above)
Comment (vicky17d) [inline src/main/java/implementations2/MyBigInteger.java]: Fixed.

File: src/main/java/implementations2/MyBigInteger.java (hunk 1 above)
Comment (vicky17d) [inline src/main/java/implementations2/MyBigInteger.java]: Fixed

File: src/main/java/implementations2/MyBigInteger.java (hunk 4 above)
Comment (vicky17d) [inline src/main/java/implementations2/MyBigInteger.java]: Fixed
    public int hashCode() {
        char sign = getSign();
        String unsignedString = getUnsignedString(inputString);
        return ((int)sign)*31 + unsignedString.hashCode();
    }

File: src/main/java/implementations2/MyBigInteger.java (hunk 5 above)
Comment (vicky17d) [inline src/main/java/implementations2/MyBigInteger.java]: Fixed.

File: src/main/java/implementations2/MyBigInteger.java (hunk 6 above)
Comment (vicky17d) [inline src/main/java/implementations2/MyBigInteger.java]: Fixed.

File: src/main/java/implementations2/MyBigInteger.java (hunk 7 above)
Comment (vicky17d) [inline src/main/java/implementations2/MyBigInteger.java]: Fixed.  throwing IllegalStateException now

File: src/main/java/implementations2/MyBigInteger.java (hunk 8 above)
Comment (vicky17d) [inline src/main/java/implementations2/MyBigInteger.java]: Fixed.

File: src/main/java/implementations2/MyBigInteger.java (hunk 9 above)
Comment (vicky17d) [inline src/main/java/implementations2/MyBigInteger.java]: Good point. Fixed.

File: src/test/java/implementations2/MyBigIntegerTest.java (hunk 16 above)
Comment (vicky17d) [inline src/test/java/implementations2/MyBigIntegerTest.java]: Makes sense. Converted the fields to variables in corresponding tests. (Fixed)

File: src/test/java/implementations2/MyBigIntegerTest.java (hunk 19 above)
Comment (vicky17d) [inline src/test/java/implementations2/MyBigIntegerTest.java]: Good point. Fixed in separate tests but may be I can combine them too
@Test
    public void testEquals() throws Exception {
        MyBigInteger big1 = new MyBigInteger("123");
        MyBigInteger big2 = new MyBigInteger("+123");
        MyBigInteger big3 = new MyBigInteger("-123");

        MyBigInteger big4 = big2;

        //Reflexive Check
        assertTrue(big1.equals(big1));
        //Symmtery check
        assertTrue(big1.equals(big2) && big2.equals(big1));
        //Transitive Check
        assertTrue(big1.equals(big4));
}

    @Test
    public void testHashCode() throws Exception {
        MyBigInteger big1 = new MyBigInteger("123");
        MyBigInteger big2 = new MyBigInteger("+123");
        MyBigInteger big3 = new MyBigInteger("-123");

        MyBigInteger big4 = big2;

        assertTrue(big1.hashCode() == big1.hashCode());
        assertTrue(big1.hashCode() == big2.hashCode());
        assertTrue(big1.hashCode() == big4.hashCode());
        assertFalse(big1.hashCode() == big3.hashCode());
}
